### ISSUE TITLE
DNS issue when adding/removing services

### ISSUE THREAD

**Issue by deadbeef84** [rancher/rancher#4284]
Rancher Version: 1.0.0

Docker Version: 1.10.3

OS: Debian 8.3 Jessie (host), alpine 3.3 (container)

Steps to Reproduce:
- Create stack `dnstest` with 2 services:

```
dns-test-alpine:
  command:
  - node
  - -e
  - 'function t() { var time = process.hrtime(); return time[0] * 1000 + time[1] /
    1e6; } function a() { var  s = t(); require("dns").lookup("whatever.dnstest.rancher.internal",
    (err, addresses, family) => { err ? console.error(err, t() - s) : console.log(addresses,
    family, t() - s); setTimeout(a, 100) }) }; a()'
  image: mhart/alpine-node:5.10.0
  links:
  - 'whatever:'
whatever:
  image: httpd
```
- The first service will do a dns-lookup every 100ms
- While the stack is running, **add another service to the stack**

Results: 

```
dns-test-alpine_2 | 2016-04-04T13:51:49.200443701Z { [Error: getaddrinfo EAI_AGAIN]
dns-test-alpine_2 | 2016-04-04T13:51:49.200501823Z   code: 'EAI_AGAIN',
dns-test-alpine_2 | 2016-04-04T13:51:49.200513448Z   errno: 'EAI_AGAIN',
dns-test-alpine_2 | 2016-04-04T13:51:49.200522747Z   syscall: 'getaddrinfo',
dns-test-alpine_2 | 2016-04-04T13:51:49.200531006Z   hostname: 'whatever.dnstest.rancher.internal' } 5002.109681010246
```

I tried the same thing using a debian/jessie image, got a different error:
`Error: getaddrinfo ENOTFOUND whatever.dnstest.rancher.internal`

Expected: Successful DNS lookup.

**Comment by deadbeef84** [comment]
Possibly related to #1699?

**Comment by deadbeef84** [comment]
Here's the contents of **rancher-dns.log**:
```
time="2016-04-04T13:51:36Z" level=info msg="Loaded answers" 
time="2016-04-04T13:51:37Z" level=info msg="Writing pid 24869 to /var/run/rancher-dns.pid" 
time="2016-04-04T13:51:37Z" level=info msg="Starting rancher-dns v0.9.1" 
time="2016-04-04T13:51:37Z" level=info msg="Loaded answers" 
time="2016-04-04T13:51:37Z" level=info msg="Listening for Reload on 127.0.0.1:8113" 
time="2016-04-04T13:51:37Z" level=info msg="Listening on :53" 
time="2016-04-04T13:51:46Z" level=info msg="Writing pid 24948 to /var/run/rancher-dns.pid" 
time="2016-04-04T13:51:46Z" level=info msg="Starting rancher-dns v0.9.1" 
time="2016-04-04T13:51:46Z" level=info msg="Loaded answers" 
time="2016-04-04T13:51:46Z" level=info msg="Listening for Reload on 127.0.0.1:8113" 
time="2016-04-04T13:51:46Z" level=info msg="Listening on :53" 
time="2016-04-04T13:51:48Z" level=info msg="Writing pid 24974 to /var/run/rancher-dns.pid" 
time="2016-04-04T13:51:48Z" level=info msg="Starting rancher-dns v0.9.1" 
time="2016-04-04T13:51:48Z" level=info msg="Loaded answers" 
time="2016-04-04T13:51:48Z" level=info msg="Listening for Reload on 127.0.0.1:8113" 
time="2016-04-04T13:51:48Z" level=info msg="Listening on :53" 
time="2016-04-04T13:51:54Z" level=info msg="Writing pid 24989 to /var/run/rancher-dns.pid" 
time="2016-04-04T13:51:54Z" level=info msg="Starting rancher-dns v0.9.1" 
time="2016-04-04T13:51:54Z" level=info msg="Loaded answers" 
time="2016-04-04T13:51:54Z" level=info msg="Listening for Reload on 127.0.0.1:8113" 
time="2016-04-04T13:51:54Z" level=info msg="Listening on :53" 
```
I had 4 hosts running, and the service set to run on all hosts, which I guess is also why we see the rancher-dns reloading 4 times.

**Comment by ronag** [comment]
This is a quite serious issue. If you have a service that will crash on `EAI_AGAIN` causing rancher to re-create it, causing `EAI_AGAIN` again, which crashes the service again... you get an infinite loop...

**Comment by vincent99** [comment]
Is there anything in the `monit` log?

**Comment by deadbeef84** [comment]
@vincent99 here are the logs for both services when reproducing the error:

**monit.log**
```
[UTC Apr  8 09:59:46] info     : Awakened by the SIGHUP signal
[UTC Apr  8 09:59:46] info     : Reinitializing monit - Control file '/etc/monit/monitrc'
[UTC Apr  8 09:59:46] info     : Shutting down monit HTTP server
[UTC Apr  8 09:59:47] info     : monit HTTP server stopped
[UTC Apr  8 09:59:47] info     : Starting monit HTTP server at [localhost:2812]
[UTC Apr  8 09:59:47] info     : monit HTTP server started
[UTC Apr  8 09:59:47] info     : '4294fc7968a7' Monit reloaded
[UTC Apr  8 10:00:57] error    : 'rancher-dns' process is not running
[UTC Apr  8 10:00:57] info     : 'rancher-dns' trying to restart
[UTC Apr  8 10:00:57] info     : 'rancher-dns' start: /etc/init.d/rancher-dns
[UTC Apr  8 10:01:02] info     : 'rancher-dns' process is running with pid 10292
```

**racher-dns.log**
```
time="2016-04-08T09:59:45Z" level=info msg="Loaded answers" 
time="2016-04-08T09:59:46Z" level=info msg="Loaded answers" 
time="2016-04-08T10:00:04Z" level=info msg="Loaded answers" 
time="2016-04-08T10:00:57Z" level=info msg="Writing pid 10292 to /var/run/rancher-dns.pid" 
time="2016-04-08T10:00:57Z" level=info msg="Starting rancher-dns v0.9.1" 
time="2016-04-08T10:00:57Z" level=info msg="Loaded answers" 
time="2016-04-08T10:00:57Z" level=info msg="Listening for Reload on 127.0.0.1:8113" 
time="2016-04-08T10:00:57Z" level=info msg="Listening on :53" 
```

**Comment by deadbeef84** [comment]
Managed to get stderr from rancher-dns:
```
fatal error: concurrent map writes

goroutine 331 [running]:
runtime.throw(0x9d7600, 0x15)
	/usr/local/go/src/runtime/panic.go:530 +0x90 fp=0xc8204eb820 sp=0xc8204eb808
runtime.mapassign1(0x84c840, 0xc8203a8d50, 0xc8204eb928, 0xc8204eb910)
	/usr/local/go/src/runtime/hashmap.go:445 +0xb1 fp=0xc8204eb8c8 sp=0xc8204eb820
main.clientSpecificCacheHit(0xc8203c3c20, 0xc, 0xc82048e5a0, 0x1)
	/go/src/github.com/rancher/rancher-dns/cache.go:21 +0x230 fp=0xc8204eb968 sp=0xc8204eb8c8
main.route(0x7fa99c7b5220, 0xc8203bce70, 0xc82048e5a0)
	/go/src/github.com/rancher/rancher-dns/main.go:211 +0x1199 fp=0xc8204ebce0 sp=0xc8204eb968
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.HandlerFunc.ServeDNS(0xa4cba0, 0x7fa99c7b5220, 0xc8203bce70, 0xc82048e5a0)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:82 +0x3a fp=0xc8204ebd00 sp=0xc8204ebce0
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*ServeMux).ServeDNS(0xc8200148a0, 0x7fa99c7b5220, 0xc8203bce70, 0xc82048e5a0)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:186 +0xbb fp=0xc8204ebd98 sp=0xc8204ebd00
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*Server).serve(0xc820065ad0, 0x7fa99f9d8bd8, 0xc8203a8e40, 0x7fa99f9d8b88, 0xc8200148a0, 0xc8200e7800, 0x2f, 0x200, 0xc82002a438, 0xc8203c3b60, ...)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:535 +0x7c8 fp=0xc8204ebf48 sp=0xc8204ebd98
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1998 +0x1 fp=0xc8204ebf50 sp=0xc8204ebf48
created by github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*Server).serveUDP
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:489 +0x3d5

goroutine 1 [IO wait]:
net.runtime_pollWait(0x7fa99f9d8a60, 0x72, 0x0)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc82012ac30, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82012ac30, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(0xc82012abd0, 0x0, 0x7fa99f9d8bb0, 0xc820136de0)
	/usr/local/go/src/net/fd_unix.go:426 +0x27c
net.(*TCPListener).AcceptTCP(0xc82002a428, 0x8db440, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:254 +0x4d
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*Server).serveTCP(0xc820065ba0, 0xc82002a428, 0x0, 0x0)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:435 +0x1c2
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*Server).ListenAndServe(0xc820065ba0, 0x0, 0x0)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:309 +0x3ec
main.main()
	/go/src/github.com/rancher/rancher-dns/main.go:76 +0x8c0

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1998 +0x1

goroutine 5 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:116 +0x132
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 6 [select, locked to thread]:
runtime.gopark(0xa4d2f8, 0xc820028728, 0x98d3f0, 0x6, 0x18, 0x2)
	/usr/local/go/src/runtime/proc.go:262 +0x163
runtime.selectgoImpl(0xc820028728, 0x0, 0x18)
	/usr/local/go/src/runtime/select.go:392 +0xa67
runtime.selectgo(0xc820028728)
	/usr/local/go/src/runtime/select.go:215 +0x12
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal1_unix.go:279 +0x358
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1998 +0x1

goroutine 7 [chan receive]:
main.watchSignals.func1(0xc8200624e0)
	/go/src/github.com/rancher/rancher-dns/main.go:121 +0x64
created by main.watchSignals
	/go/src/github.com/rancher/rancher-dns/main.go:125 +0x13f

goroutine 8 [chan receive]:
main.watchSignals.func2()
	/go/src/github.com/rancher/rancher-dns/main.go:128 +0x50
created by main.watchSignals
	/go/src/github.com/rancher/rancher-dns/main.go:134 +0x157

goroutine 9 [IO wait]:
net.runtime_pollWait(0x7fa99f9d88e0, 0x72, 0x0)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc820166060, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc820166060, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(0xc820166000, 0x0, 0x7fa99f9d8bb0, 0xc8202d1800)
	/usr/local/go/src/net/fd_unix.go:426 +0x27c
net.(*TCPListener).AcceptTCP(0xc82016a000, 0x463340, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:254 +0x4d
net/http.tcpKeepAliveListener.Accept(0xc82016a000, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:2427 +0x41
net/http.(*Server).Serve(0xc820160000, 0x7fa99d18e028, 0xc82016a000, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:2117 +0x129
net/http.(*Server).ListenAndServe(0xc820160000, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:2098 +0x136
net/http.ListenAndServe(0x9907a0, 0xe, 0x7fa99f9d7a98, 0xc820012820, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:2195 +0x98
created by main.watchHttp
	/go/src/github.com/rancher/rancher-dns/main.go:142 +0x301

goroutine 10 [IO wait]:
net.runtime_pollWait(0x7fa99f9d89a0, 0x72, 0x70)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc82012aca0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82012aca0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).readMsg(0xc82012ac40, 0xc8200e7a00, 0x200, 0x200, 0xc8203a8e70, 0x28, 0x28, 0xffffffffffffffff, 0x0, 0x0, ...)
	/usr/local/go/src/net/fd_unix.go:304 +0x33d
net.(*UDPConn).ReadMsgUDP(0xc82002a438, 0xc8200e7a00, 0x200, 0x200, 0xc8203a8e70, 0x28, 0x28, 0x200, 0x84eac0, 0x1, ...)
	/usr/local/go/src/net/udpsock_posix.go:96 +0x171
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.ReadFromSessionUDP(0xc82002a438, 0xc8200e7a00, 0x200, 0x200, 0x200, 0x200, 0x0, 0x0)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/udp.go:47 +0xc1
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*Server).readUDP(0xc820065ad0, 0xc82002a438, 0x77359400, 0x0, 0x0, 0x0, 0xc82003bd90, 0x0, 0x0)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:605 +0x190
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*defaultReader).ReadUDP(0xc82002a460, 0xc82002a438, 0x77359400, 0x0, 0x0, 0x0, 0xc8200148a0, 0x0, 0x0)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:230 +0x5f
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*Server).serveUDP(0xc820065ad0, 0xc82002a438, 0x0, 0x0)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:478 +0x1fe
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*Server).ListenAndServe(0xc820065ad0, 0x0, 0x0)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:327 +0x6fc
main.main.func1(0xc820065ad0)
	/go/src/github.com/rancher/rancher-dns/main.go:73 +0x25
created by main.main
	/go/src/github.com/rancher/rancher-dns/main.go:74 +0x751

goroutine 330 [runnable]:
github.com/rancher/rancher-dns/vendor/github.com/Sirupsen/logrus.WithFields(0xc820473dd0, 0xc820473dd0)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/Sirupsen/logrus/exported.go:66
main.route(0x7fa99c7b5220, 0xc82021f650, 0xc8203c8000)
	/go/src/github.com/rancher/rancher-dns/main.go:227 +0x1cc1
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.HandlerFunc.ServeDNS(0xa4cba0, 0x7fa99c7b5220, 0xc82021f650, 0xc8203c8000)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:82 +0x3a
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*ServeMux).ServeDNS(0xc8200148a0, 0x7fa99c7b5220, 0xc82021f650, 0xc8203c8000)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:186 +0xbb
github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*Server).serve(0xc820065ad0, 0x7fa99f9d8bd8, 0xc8203a8de0, 0x7fa99f9d8b88, 0xc8200148a0, 0xc820461c00, 0x2f, 0x200, 0xc82002a438, 0xc8203c3b40, ...)
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:535 +0x7c8
created by github.com/rancher/rancher-dns/vendor/github.com/miekg/dns.(*Server).serveUDP
	/go/src/github.com/rancher/rancher-dns/vendor/github.com/miekg/dns/server.go:489 +0x3d5
```

**Comment by ronag** [comment]
@vincent99: `fatal error: concurrent map writes`?

**Comment by joshwget** [comment]
Should be fixed with rancher/rancher-dns#31

**Comment by joshwget** [comment]
If anyone continues to experience this, please open a new issue.
